[Writing Tools] Tapping on Restart after a Magic Rewrite duplicates the sentence

rr-codes opened the pull request:
#### 405782ff2e808caa3d518d03fc5de7b15de8098b
<pre>
[Writing Tools] Tapping on Restart after a Magic Rewrite duplicates the sentence
<a href="https://bugs.webkit.org/show_bug.cgi?id=275818">https://bugs.webkit.org/show_bug.cgi?id=275818</a>
<a href="https://rdar.apple.com/130065512">rdar://130065512</a>

Reviewed by Aditya Keerthi.

In the existing implementation of Rewrite/Composition, there were two fundamental issues:

* The stateful context range must always be up-to-date, which was a fragile architecture as it relied
upon it being updated with every different type of replacement operation.

* While undo/redo happened to work for trivial cases, the WritingToolsController was not aware of these
changes happening, which would also put the context range out of date. Additionally, each individual ReplaceSelectionCommand
was being undone or redone at a time, instead of the entire set of constituent commands belong to a particular composition.

To fix this, introduce a new specialized type of edit command specifically designed for rewrites and compositions,
and have it encapsulate both a group of ReplaceSelectionCommands as well as its corresponding context range.
This facilitates the ability to undo/redo a group of commands at once, and also ensures that the context range is always correct
for a given command.

To ensure that the right command is always being used, an undo and redo stack are now used to store the commands,
and a given command moves from one stack to the other when it is undone or redone. This also allows the &quot;show original&quot;
and &quot;show rewritten&quot; commands to function by simply undo-ing or redo-ing all the commands.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::unappliedEditing):
(WebCore::Editor::reappliedEditing):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::rangeExpandedAroundRangeByCharacters):
* Source/WebCore/editing/VisibleUnits.h:
* Source/WebCore/editing/WritingToolsCompositionCommand.cpp: Added.
(WebCore::WritingToolsCompositionCommand::WritingToolsCompositionCommand):
(WebCore::WritingToolsCompositionCommand::replaceFragment):
(WebCore::WritingToolsCompositionCommand::doApply):
* Source/WebCore/editing/WritingToolsCompositionCommand.h: Added.
(WebCore::WritingToolsCompositionCommand::create):
(WebCore::WritingToolsCompositionCommand::endingContextRange const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::respondToUnappliedWritingToolsEditing):
(WebCore::Page::respondToReappliedWritingToolsEditing):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::WritingToolsController):
(WebCore::WritingToolsController::willBeginWritingToolsSession):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
(WebCore::WritingToolsController::writingToolsSessionDidReceiveAction&lt;WritingTools::Session::Type::Composition&gt;):
(WebCore::WritingToolsController::respondToUnappliedEditing):
(WebCore::WritingToolsController::respondToReappliedEditing):
(WebCore::WritingToolsController::contextRangeForSessionWithID const):
(WebCore::WritingToolsController::replaceContentsOfRangeInSession):
(WebCore::WritingToolsController::replaceContentsOfRangeInSessionInternal): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, CompositionWithUndo)):
(TEST(WritingTools, CompositionWithMultipleUndosAndRestarts)):
(TEST(WritingTools, CompositionWithUndoAndRestart)):
(TEST(WritingTools, Composition)):
(TEST(WritingTools, CompositionRevert)):
(TEST(WritingTools, CompositionWithMultipleChunks)):
(TEST(WritingTools, ContextRangeFromCaretSelection)):
(TEST(WritingTools, ContextRangeFromRangeSelection)):

Canonical link: <a href="https://commits.webkit.org/280437@main">https://commits.webkit.org/280437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14f2b5d7f3ba032df95caa246d8e07b425ad0004

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60236 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7254 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4944 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58652 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48869 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/30582 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6200 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6066 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6471 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61919 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/534 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6578 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53131 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48927 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/461 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8418 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31777 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32863 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33947 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32609 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -463,39 +461,47 @@
+    // Smart reply composition sessions never send any actions to the delegate.
+    ASSERT(session.compositionType != WritingTools::Session::CompositionType::SmartReply);
+
+    RefPtr document = this->document();
```
is this needed?

rr-codes replied:
yes

aprotyas commented on Source/WebCore/editing/WritingToolsCompositionCommand.h:
```diff
@@ -0,0 +1,67 @@
+private:
+    WritingToolsCompositionCommand(Ref<Document>&&, const SimpleRange&);
+
+    void doApply() override;
```
Since it's trivial, maybe just this?
```suggestion
    void doApply() override { }
```

aprotyas commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -632,6 +640,42 @@
+        if (!state)
+            continue;
+
+        auto appliedCommandIsWritingToolsCommand = std::ranges::any_of(state->reappliedCommands, [&](const auto& command) {
```
Nit: just capture `composition`

aprotyas commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -632,6 +640,42 @@
+        if (!state)
+            continue;
+
+        auto appliedCommandIsWritingToolsCommand = std::ranges::any_of(state->unappliedCommands, [&](const auto& command) {
```
I think we can make a reusable helper for this?

```
static bool appliedCommandIsWritingToolsCommand(const auto& commands, EditCommandComposition* composition)
{
    return std::ranges::any_of(commands, [composition](const auto& command) {
        return &command->ensureComposition() == composition;
    });
}
```

pxlcoder replied:
+1 if a method on the object itself is not possible.

aprotyas reviewed (commented):
This looks good to land for me! Though someone more familiar should review the changes under `page/writing-tools/WritingToolsController`...

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -632,6 +640,42 @@
+        if (!state)
+            continue;
+
+        auto appliedCommandIsWritingToolsCommand = std::ranges::any_of(state->reappliedCommands, [&](const auto& command) {
```
Would it be possible to detect that the command is a Writing Tools command via method on `EditCommandComposition` or another command class?

I expect I'll need something similar.

pxlcoder replied:
(Chatted offline about adding a new `EditAction` for this)

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -588,6 +588,186 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
+            EXPECT_WK_STREQ(rewrittenText, [webView contentsAsStringWithoutNBSP]);
+        };
+
+        rewrite(@"A");
```
Nit: Maybe assert the contents of the web view after this rewrite?

rr-codes replied:
I already do, inside the block

rr-codes committed:
[Writing Tools] Tapping on Restart after a Magic Rewrite duplicates the sentence
https://bugs.webkit.org/show_bug.cgi?id=275818
rdar://130065512

Reviewed by Aditya Keerthi.

In the existing implementation of Rewrite/Composition, there were two fundamental issues:

* The stateful context range must always be up-to-date, which was a fragile architecture as it relied
upon it being updated with every different type of replacement operation.

* While undo/redo happened to work for trivial cases, the WritingToolsController was not aware of these
changes happening, which would also put the context range out of date. Additionally, each individual ReplaceSelectionCommand
was being undone or redone at a time, instead of the entire set of constituent commands belong to a particular composition.

To fix this, introduce a new specialized type of edit command specifically designed for rewrites and compositions,
and have it encapsulate both a group of ReplaceSelectionCommands as well as its corresponding context range.
This facilitates the ability to undo/redo a group of commands at once, and also ensures that the context range is always correct
for a given command.

To ensure that the right command is always being used, an undo and redo stack are now used to store the commands,
and a given command moves from one stack to the other when it is undone or redone. This also allows the "show original"
and "show rewritten" commands to function by simply undo-ing or redo-ing all the commands.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::unappliedEditing):
(WebCore::Editor::reappliedEditing):
* Source/WebCore/editing/VisibleUnits.cpp:
(WebCore::rangeExpandedAroundRangeByCharacters):
* Source/WebCore/editing/VisibleUnits.h:
* Source/WebCore/editing/WritingToolsCompositionCommand.cpp: Added.
(WebCore::WritingToolsCompositionCommand::WritingToolsCompositionCommand):
(WebCore::WritingToolsCompositionCommand::replaceFragment):
(WebCore::WritingToolsCompositionCommand::doApply):
* Source/WebCore/editing/WritingToolsCompositionCommand.h: Added.
(WebCore::WritingToolsCompositionCommand::create):
(WebCore::WritingToolsCompositionCommand::endingContextRange const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::respondToUnappliedWritingToolsEditing):
(WebCore::Page::respondToReappliedWritingToolsEditing):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::WritingToolsController):
(WebCore::WritingToolsController::willBeginWritingToolsSession):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
(WebCore::WritingToolsController::writingToolsSessionDidReceiveAction<WritingTools::Session::Type::Composition>):
(WebCore::WritingToolsController::respondToUnappliedEditing):
(WebCore::WritingToolsController::respondToReappliedEditing):
(WebCore::WritingToolsController::contextRangeForSessionWithID const):
(WebCore::WritingToolsController::replaceContentsOfRangeInSession):
(WebCore::WritingToolsController::replaceContentsOfRangeInSessionInternal): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, CompositionWithUndo)):
(TEST(WritingTools, CompositionWithMultipleUndosAndRestarts)):
(TEST(WritingTools, CompositionWithUndoAndRestart)):
(TEST(WritingTools, Composition)):
(TEST(WritingTools, CompositionRevert)):
(TEST(WritingTools, CompositionWithMultipleChunks)):
(TEST(WritingTools, ContextRangeFromCaretSelection)):
(TEST(WritingTools, ContextRangeFromRangeSelection)):

Canonical link: https://commits.webkit.org/280437@main
